### ISSUE TITLE
using same stake amount for buy and sell

### ISSUE THREAD

**Issue by JanAbraham** [freqtrade/freqtrade#2639]
Currently, sell orders are triggered with the amount of assets which have been bought. Is it possible to have an option in the strategy which decides whether the sell order is having the same stake amount or the same asset amount as the buy order?

The logic for associated stop-loss orders and trailings can remain the same.

Unfortunately, I am not a developer, so I cannot contribute any code proposals here. Still hope for your support on that one :-)

Cheers,
Jan

**Comment by hroff-1902** [comment]
is it about rebalancing with equal amounts?

**Comment by xmatthias** [comment]
If i understand this correctly, then it would leave the gains in the assets it bought, and pull out the invested stake_currency only (so you buy 50USDT - and sell 50USDT - keeping the 0.5 XRP gain in XRP).

I don't think that makes much sense (at least not without explanation why you think this should be useful / of the usecase) - it'll end up with accounts owning 100'ds of currencies, with probably very low value (unless one bumps ...).

Playing your strategy further (at least how i understood it) shows that this approach will fail:

You'd leave profits in the asset you bought (so you have 20 assets with a low value) - however when you sell at a loss, you'd sell 48 instead of the 50 USDT you invested. 
Since you're never pulling out gains from the trades, the bot would  run out of stake_currency, since you will have an eventual stoploss / trade moving in the wrong direction.
The quality of the strategy  would only determine the duration it takes to run out of stake_currency - but it would happen sooner or later (there'll always be some stoplosses ... being prepared to take the (precalculated) loss is part of being a trader).

Since this would mean that the bot can only run for X amount of time before it's guaranteed to run out of stake_currency, i don't think this should be even an option in freqtrade.

**Comment by JanAbraham** [comment]
Thanks for sharing your thoughts, Matthias! Highly appreciate that!

This trading strategy is a long-term investment strategy. Each investment is based on the value of an asset. This strategy builds up assets over time and utilises volatility of a relatively young marked (cryptocurrencies). However, it is a known strategy with share trading from the good old days.

Yes, stake-currency is running out over time. I would rephrase it into: stake-currency is being converted into assets. It can either be added-up by another investment or by selling collected assets.

How many different assets / currencies the bot collects depends on how many assets a trader decides to invest into. If gambling is in ones mind and jumping opportunistically from asset to asset then I absolutely agree with you. After a little homework, I came to only about 20 coins which found an underlying value/purpose/idea which I like to invest into.

**Comment by JanAbraham** [comment]
> is it about rebalancing with equal amounts?

Not sure what you mean with rebalancing, @hroff-1902

**Comment by xmatthias** [comment]
what you're suggesting is more a "holding" strategy than a trading strategy (or more portfolio management). 
Yes, it'll rebalance assets from time to time - but i don't think freqtrade is the right tool for that (nor does it aim to be ... ).

The way freqtrade works is that you have one currency (`stake_currency`) - and execute trades based on that with the aim to increase the amount `stake_currency` via a (hopefully) profitable strategy.

Even assuming we'd support multiple stake-currencies in the future there's a ton or problems.

Technically, #2183 may accidentally add what you ask ... but that's not going to be the end of what you'll be needing to succeed. 

You'll need to trade any asset with any other asset available on the exchange ... 
which will get you blocked from the exchange with freqtrade in no time - since you'd need to add every possible pair (if you say 20 pairs - that's 20 x 20 pairs) to the whitelist, which is then queried from the exchange at least once per candle (even assuming daily candles, it's a bunch of requests, probably with the result of the latter pairs not getting any results back due to exchange blocks).

to make things more difficult, not all combinations will be available - so it'll need to support to trade X to Y - or X to Y over Z, depending on availability.

If you only trade against `stake_currency` (the way freqtrade works), then a rebalance (say from BTC to ETH - using USDT as stake currency) will cost you twice the fees you should be spending (once to get out of BTC, once to get into ETH).

simply put - what you'd need is something else - freqtrade would need to be almost completely rewritten from scratch to make this strategy work without wasting money.

**Comment by hroff-1902** [comment]
> > is it about rebalancing with equal amounts?
> 
> Not sure what you mean with rebalancing, @hroff-1902


* _"Portfolio rebalancing is a strategy that has been used by investors in the traditional investment markets for decades. However, this simple investment technique has been slow to integrate as a standard for cryptocurrency portfolios. In the following sections, rebalancing will be examined in critical detail. Portfolio rebalancing is the act of realigning assets in a portfolio back to their target allocations. As the market moves, assets in a portfolio will deviate from their desired percent allocation. Returning to the target allocations requires assets to be bought and sold. This process is called rebalancing. The target allocation weightings for each asset in a portfolio is determined by the investor when they are constructing a portfolio. These weightings are simply the percent of each asset that should be represented as part of the total value of the combined portfolio. In the case of cryptocurrencies, an asset can be defined as any individual coin or token. When it is time to rebalance, the assets are traded such that the value held in each asset is once again equal to the percentages that were originally specified. A practical example of this strategy would be if there was a portfolio of 4 different cryptocurrencies. After studying the market, the investor determined that the optimal allocations for each asset in their portfolio was an equal 25% stake in their portfolio. This would mean at the end of each rebalance, the portfolio would consist of 25% in each of the 4 assets. The value of each asset should be converted to a single common value, such as BTC or a fiat currency, so the different assets won’t be evaluated based on quantity, but value. Therefore, if you had $100 total between these four assets, you would have $25 in each after a rebalance took place."_

**Comment by JanAbraham** [comment]
I think it is not that complex as you describe (I am doing this strategy since > 1 year with Tradingview/Autoview). Since I am missing some functionalities there, which freqtrade offers, I ended up here :-)

Let me understand your concern: What exactly is the challenge with triggering a sale trade with the exact same amount of stake_currency with which the buy trade was triggered? Essentially, "stake_amount" is the same for buy and sell in case of ROI > 0. Am I missing something?

**Comment by JanAbraham** [comment]
@hroff-1902 : Ah yes, now I know what you mean. But, no, that is not what I am talking about (even though it would be a nice function to have ;-). I am only talking about having the same stake-amount for a buy and sale trade.

**Comment by hroff-1902** [comment]
trying to understand better the point...
now stake_amount is the quote currency amount. Are you talking about the possibility to have stake_amount as the amount of the base currency? (and thus probably trade pairs BTC/USDT, BTC/ETH, i.e. with the same base currency, if these pairs are on the exchange) instead of current pairs BTC/USDT, ETH/USDT -- pairs with common quote currency?

**Comment by JanAbraham** [comment]
I see your point, but that's not mine. Let me make an example.

Order of events currently:
1. Buy order is triggered for $50 and X coins are being bought
2. Sell order is triggered and X coins are being sold again
3. Profit in $ is being made from the price-difference of the coins

My approach:
1. Buy order is triggered for $50 and X coins are being bought
2. Sell order is triggered and $50 worth of coins are being sold
3. Profit in #coins is being made, which are kept in the wallet

In case of stop-loss, the current logic applies (X coins are being sold for a lower price than $50).

**Comment by hroff-1902** [comment]
it will leave the position open on the exchange with some amount of coins bought previously...

**Comment by JanAbraham** [comment]
Not really, because we sell less coins than we have bought.

**Comment by hroff-1902** [comment]
* it will leave the position open till the next buy signal for this pair
* why should it buy less amount of coins, if you want to buy the same amount of stake_currency? then they will be sold and if the trade is profitable -- it will leave the position open again, with bigger amount of coins left.
* how to deal with stoploss -- should it sell all amount of coins, to the zero position? or what?
* how to deal with minimal_roi reached (take profit)? Shall it also sell the full position? or what?

It simply does not fit into how freqtrade operates. In your proposal you left you profit in the base coins as the remainder of position, freqtrade now returns the profit back to the quote (stake) currency and closes the position. That's the only difference. Imho, I do not see any trading benefits.

**Comment by JanAbraham** [comment]
Not sure, I understand correctly, but I am trying: Why should a position still be open? The buy trade is closed and the sell trade gets closed as well.

in case of ROI:
- it sells less coins than have been bought. The remaining coins are being kept in the wallet. Sale closed.
in case of stoploss:
- it sells the same amount of coins which have been bought. Sale closed.

Not sure if you are familiar with Autoview, but the coding there would be like that:
Traditional-$-collecting-Approach:
e=Binance a=autoview s=ENJUSDT b=buy q=600 t=market u=contracts
e=Binance a=autoview s=ENJUSDT b=sell q=600 p=2% u=contracts

Asset-collection-Approach:
e=Binance a=autoview s=ENJUSDT b=buy q=50 t=market u=currency
e=Binance a=autoview s=ENJUSDT b=sell q=50 p=2% u=currency

The trading benefit is in the collection of assets (coins), rather than $. I know it is a different concept than the standard approach, but it is working and well used in the traditional markets. Hope you can help me with that, since this is a function which would open up freqtrade to total different types of investors.

**Comment by hroff-1902** [comment]
> Why should a position still be open?

Assume you are trading pair XYZ/BTC, the price is 0.000100 (BTC per XYZ), stake amount (for opening position/trade) is 0.05 BTC. The buys 0.05/0.000100 = 500 XYZ coins. Assume you are lucky and the price is now 0.000125. You sell the coins for 0.05 BTC, i.e. 0.05/0.000125 = 400 XYZ coins. Position is still open with 100 XYZ coins in it.

As I said, I do not see any trading benefits in it. You simply propose to leave your profit in the XYZ coins instead of closing the position and transferring all 500 XYZ coins, including profit, back to BTC.

**Comment by hroff-1902** [comment]
It simply does not fit into internals of the bot to be implemented easily.

What can be the benefit in "collection of assets (coins)", rather than quote currency?

**Comment by JanAbraham** [comment]
Yes, you are right, the position is not yet closed (I mixed it up with having an open trade, which is not the case). Indeed, this is the typical behaviour with buy-and-hold investors rather with day-traders or round-trip traders. I absolutely agree with you.

The benefit is long term: One ground rule of investing is "Assets are future income." The coins I am collecting now, will lead to a (hopeful) profit in the future in case the value increases. Which is essentially the same as mining, but without investing in hardware ;-)

**Comment by hroff-1902** [comment]
As @xmatthias said above, you'll be out of quote currency (stake currency) soon or later, since you leave profits in base currencies and transfer losses back to quote currency...

**Comment by hroff-1902** [comment]
Have you tried to collect CryptoKitties? https://www.cryptokitties.co
🤣

**Comment by JanAbraham** [comment]
> As @xmatthias said above, you'll be out of quote currency (stake currency) soon or later, since you leave profits in base currencies and transfer losses back to quote currency...

Sure. Thats part of long-term-investing. Transferring fiat into assets for future income. We all do that when we buy a house and rent it out.

Looks like I need to keep my Tradingview/Autoview stuff running for the coin collection and long term investments. But I still have some use-cases for freqtrade to satisfy my day-trading fever :-) And especially the backtesting, hyperopt and edge are really cool. I have not seen them before on that level in any other tool. Well done guys and keep developing them please!

So, in case you don't see an added value of my requested function in freqtrade then you can close this case.

Only one comment more: In the quite close future the classic shares will be tokenised and trading will be available for everybody like with cryptos. People which trade those, have a different mindset. You might not want to leave them out of potentially using freqtrade ;-)

**Comment by xmatthias** [comment]
> So, in case you don't see an added value of my requested function in freqtrade then you can close this case.

I do belive it can be profiable, but I don't think that's the point - this strategy is a different type of investing.

I think freqtrade is good at what it does. While we may leave this type of investors / traders out - that's not necessarily a bad thing. 
It's better to focus on our niche (short-term trading) and do that well other than trying to be below average doing everything, and trying to help you invest in houses too because we found a online house broker who let's me buy in BTC...(extreme example, but ... :laughing: )

Trying to cover every possible trading strategy / method would create an impossible to maintain codebase, with too many options even for ourselfs to configure the bot correctly, letalone for a new user. Keep in mind that we're not working full-time on this bot (where we can hire an additional head if we need it), but it's a hobby / side project - so we need to keep it managable with the limited time we have available.

**Comment by hroff-1902** [comment]
Marked it as enhancement, now the proposal is clear. We can leave it open and hang as a proposed enhancement, but I agree with @xmatthias -- it's a different type of investing and it's not compatible with the niche of Freqtrade, so do not expect active actions in this directions in the observable future. Maybe further enhancements in Freqtrade make it easier to implement.

**Comment by JanAbraham** [comment]
I totally understand your points guys! Keep focused on what you do well. I like that. Appreciate the discussion around the topic. Liked it a lot :-)

**Comment by hroff-1902** [comment]
Closing this as this approach has very tight niche and do not fit to the freqtrade main aims as well as to the internal design solutions made in the bot.

One will probably be able to implement it as a custom set of position management rules in the future when a generic position management mechanism will be implemented in the bot.
